File: compatibility/pg/ruby/pg_test.rb
require 'pg'

class PGTest
  class Tests
    def initialize
      @conn = nil
      @tests = []
    end

    def connect(ip, port, user, password)
      begin
        @conn = PG.connect(host: ip, port: port, user: user, password: password, dbname: 'main')
      rescue PG::Error => e
        raise "Connection failed: #{e.message}"
      end
    end

    def disconnect
      @conn.close if @conn
    end

    def add_test(query, expected_results)
      @tests << Test.new(query, expected_results)
    end

    def run_tests
      @tests.each do |test|
        return false unless test.run(@conn)
      end
      true
    end

    def read_tests_from_file(filename)
      begin
        File.open(filename, 'r') do |file|
          query = nil
          results = []
          file.each_line do |line|
            line = line.strip
            if line.empty?
              if query
                add_test(query, results)
                query = nil
                results = []
              end
            else
              if query.nil?
                query = line
              else
                results << line.split(',')
              end
            end
          end
          add_test(query, results) if query
        end
      rescue IOError => e
        puts "File read error: #{e.message}"
        exit 1
      end
    end

    class Test
      def initialize(query, expected_results)
        @query = query
        @expected_results = expected_results
      end

      def run(conn)
        begin
          puts "Running test: #{@query}"
          result = conn.exec(@query)
          if result.ntuples == 0
            puts "Returns 0 rows"
            return @expected_results.empty?
          end
          if result.nfields != @expected_results[0].length
            puts "Expected #{@expected_results[0].length} columns, got #{result.nfields}"
            return false
          end
          result.each_with_index do |row, row_num|
            row.each_with_index do |(col_name, value), col_num|
              expected = @expected_results[row_num][col_num]
              if value != expected
                puts "Expected:\n'#{expected}'"
                puts "Result:\n'#{value}'\nRest of the results:"
                result.each_row { |r| puts r.join(',') }
                return false
              end
            end
          end
          puts "Returns #{result.ntuples} rows"
          if result.ntuples != @expected_results.length
            puts "Expected #{@expected_results.length} rows"
            return false
          end
          true
        rescue PG::Error => e
          puts e.message
          false
        end
      end
    end
  end

  def self.main(args)
    if args.length < 5
      puts "Usage: ruby PGTest.rb <ip> <port> <user> <password> <testFile>"
      exit 1
    end

    tests = Tests.new
    tests.connect(args[0], args[1].to_i, args[2], args[3])
    tests.read_tests_from_file(args[4])

    unless tests.run_tests
      tests.disconnect
      exit 1
    end
    tests.disconnect
  end
end

if __FILE__ == $0
  PGTest.main(ARGV)
end

File: compatibility/pg/perl/pg_test.pl
use strict;
use warnings;
use DBI;

package Tests;

sub new {
    my $class = shift;
    my $self = {
        conn => undef,
        tests => [],
    };
    bless $self, $class;
    return $self;
}

sub connect {
    my ($self, $ip, $port, $user, $password) = @_;
    my $dsn = "dbi:Pg:dbname=main;host=$ip;port=$port";
    $self->{conn} = DBI->connect($dsn, $user, $password, { RaiseError => 1, AutoCommit => 1 });
}

sub disconnect {
    my $self = shift;
    $self->{conn}->disconnect();
}

sub add_test {
    my ($self, $query, $expected_results) = @_;
    push @{$self->{tests}}, { query => $query, expected_results => $expected_results };
}

sub run_tests {
    my $self = shift;
    foreach my $test (@{$self->{tests}}) {
        return 0 unless $self->run_test($test);
    }
    return 1;
}

sub trim {
    my $s = shift;
    $s =~ s/^\s+|\s+$//g;
    return $s;
}

sub read_tests_from_file {
    my ($self, $filename) = @_;
    open my $fh, '<', $filename or die "Could not open file '$filename': $!";
    my $query;
    my @results;
    while (my $line = <$fh>) {
        $line = trim($line);
        if ($line eq '') {
            if ($query) {
                $self->add_test($query, [@results]);
                $query = undef;
                @results = ();
            }
        } elsif (!$query) {
            $query = $line;
        } else {
            push @results, [split /,/, $line];
        }
    }
    $self->add_test($query, [@results]) if $query;
    close $fh;
}

sub run_test {
    my ($self, $test) = @_;
    my $query = $test->{query};
    my $expected_results = $test->{expected_results};
    print "Running test: $query\n";

    my $st = $self->{conn}->prepare($query);
    $st->execute();
    my $rows = $st->rows;
    if ($rows == 0 || $rows == -1) {
        print "No rows returned\n";
        return defined $expected_results && @$expected_results == 0;
    }
    my $cols = $st->{NUM_OF_FIELDS};
    if (!defined $cols) {
        print "No columns returned\n";
        return defined $expected_results && @$expected_results == 0;
    }
    if ($cols != @{$expected_results->[0]}) {
        print "Expected " . @{$expected_results->[0]} . " columns\n";
        return 0;
    }
    my $row_num = 0;
    while (my @row = $st->fetchrow_array) {
        for my $col_num (0..$#row) {
            if ($row[$col_num] ne $expected_results->[$row_num][$col_num]) {
                print "Expected:\n'$expected_results->[$row_num][$col_num]'\n";
                print "Result:\n'$row[$col_num]'\nRest of the results:\n";
                while (my @rest = $st->fetchrow_array) {
                    print join(',', @rest) . "\n";
                }
                return 0;
            }
        }
        $row_num++;
    }
    print "Returns $row_num rows\n";
    if ($row_num != @$expected_results) {
        print "Expected " . @$expected_results . " rows\n";
        return 0;
    }
    return 1;
}

package main;

if (@ARGV < 5) {
    die "Usage: perl PGTest.pl <ip> <port> <user> <password> <testFile>\n";
}

my $tests = Tests->new();
$tests->connect($ARGV[0], $ARGV[1], $ARGV[2], $ARGV[3]);
$tests->read_tests_from_file($ARGV[4]);

if (!$tests->run_tests()) {
    $tests->disconnect();
    exit 1;
}
$tests->disconnect();